### PR TITLE
feat(frontend): send coins via identifier

### PR DESCRIPTION
**Pull request by Mogge**
## 🍰 Pullrequest

Watch out: __based on #2988__

image

### Issues

- I cannot find any issues the this topic

### Todo

- [X] None

### CODE REVIEW

File: frontend/src/components/GddSend/TransactionForm.spec.js [hunk 1]
```diff
@@ -116,51 +116,54 @@ describe('TransactionForm', () => {
...
+        describe('identifier field', () => {
+          it('has an input field of type text', () => {
             expect(
-              wrapper.find('div[data-test="input-email"]').find('input').attributes('type'),
-            ).toBe('email')
+              wrapper.find('div[data-test="input-identifier"]').find('input').attributes('type'),
+            ).toBe('text')
           })
 
           it('has a label form.receiver', () => {
```
Comment (ulfgebhardt) [inline frontend/src/components/GddSend/TransactionForm.spec.js:126]:
```suggestion
          it('has a label form.recipient', () => {
```

File: frontend/src/validation-rules.js [hunk 2]
```diff
@@ -2,6 +2,13 @@ import { configure, extend } from 'vee-validate'
 // eslint-disable-next-line camelcase
 import { required, email, min, max, is_not } from 'vee-validate/dist/rules'
 import { checkUsername } from '@/graphql/queries'
+import { validate as validateUuid, version as versionUuid } from 'uuid'
+
+// taken from vee-validate
+// eslint-disable-next-line no-useless-escape
+const emailRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
```
Comment (ulfgebhardt) [inline frontend/src/validation-rules.js:9]:
```suggestion
const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
```

File: frontend/src/validation-rules.js [hunk 3]
```diff
@@ -2,6 +2,13 @@ import { configure, extend } from 'vee-validate'
...
+
+const usernameRegex = /^(?=.{3,20}$)[a-zA-Z0-9]+(?:[_-][a-zA-Z0-9]+?)*$/
```
Comment (ulfgebhardt) [inline frontend/src/validation-rules.js:11]:
```suggestion
const USERNAME_REGEX = /^(?=.{3,20}$)[a-zA-Z0-9]+(?:[_-][a-zA-Z0-9]+?)*$/
```

File: frontend/src/validation-rules.js [hunk 4]
```diff
@@ -141,7 +148,7 @@ export const loadAllRules = (i18nCallback, apollo) => {
 
   extend('usernameUnique', {
     validate(value) {
-      if (!value.match(/^(?=.{3,20}$)[a-zA-Z0-9]+(?:[_-][a-zA-Z0-9]+?)*$/)) return true
+      if (!value.match(usernameRegex)) return true
```
Comment (ulfgebhardt) [inline frontend/src/validation-rules.js:151]:
```suggestion
      if (!value.match(USERNAME_REGEX)) return true
```

File: frontend/src/validation-rules.js [hunk 5]
```diff
@@ -155,4 +162,14 @@ export const loadAllRules = (i18nCallback, apollo) => {
     },
     message: (_, values) => i18nCallback.t('form.validation.username-unique', values),
   })
+
+  extend('validIdentifier', {
+    validate(value) {
+      const isEmail = !!emailRegex.test(value)
```
Comment (ulfgebhardt) [inline frontend/src/validation-rules.js:168]:
```suggestion
      const isEmail = !!EMAIL_REGEX.test(value)
```

File: frontend/src/validation-rules.js [hunk 6]
```diff
@@ -155,4 +162,14 @@ export const loadAllRules = (i18nCallback, apollo) => {
...
+      const isUsername = !!value.match(usernameRegex)
```
Comment (ulfgebhardt) [inline frontend/src/validation-rules.js:169]:
```suggestion
      const isUsername = !!value.match(USERNAME_REGEX)
```

**Review by ulfgebhardt — CHANGES_REQUESTED** [review]
Works very well and without problems.

I could send via
- Menu Link
- EMail
- Username
- Gradido-ID

Invalid Identifiers fail, as do invalid uuids.

image

Usability wise we have the identifier shown in the confirmation dialog, which is very use-unfriendly
image

I have some minor suggestions for code change

**Review by Elweyn — APPROVED** [review]
Works perfectly for me! Have tested send with GradidoID, set an alias, send with alias and last with the email!

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**